cumulusci/core/utils.py

```diff
@@ -26,6 +32,26 @@ def import_global(path):
 import_class = import_global
 
 
+def conditional_deprecation_warning(message):
+    "This is a no-op unless it is turned on by pytest usage or `python -Wd`"
+    assert "PYTEST_CURRENT_TEST" not in os.environ, message
+    # hidden by default
+    warnings.warn(message, DeprecationWarning)
+
+
+def deprecated_import(fully_qualified):
+    cls = import_global(fully_qualified)
+    clsname = cls.__name__
+    message = f"Instead of importing from {__name__}.{clsname}, please import from {fully_qualified}"
+
+    def __init__(self, *args, **kwargs):
+        conditional_deprecation_warning(message)
+        super(cls, self).__init__(*args, **kwargs)
+
+    init_dict = {"__init__": __init__}
+    return type(clsname, (cls,), init_dict)
```

Comment:
This looks a bit too magical. If there's an isinstance check using the class from the old location, it will fail to return true for instances of the class from the new location.

I guess the point is to defer warnings until pytest has initialized sufficiently to determine whether we're supposed to raise or warn? How about this as an alternative:
1. Warn unconditionally at import time
2. Make deprecated_import add the class to a global set, and use a pytest hook to raise assertion errors between test discovery and starting to run the tests.

Reply:
If the name bound in `__init__.py` is bound to exactly the same object as foo_bar.py then there is no way to know which one is being imported by the client file.

deprecated_import happens unconditionally as soon as `__init__.py` is imported for any reason, whether legitimate or not.

I don't know that there is a way to preserve bidirectional isinstance and also issue the deprecation warnings only when they are really merited. Bidirectional isinstance is probably (?) the same as "is" 

if we deprecate every single class in `__init__.py` maybe we could deprecate importing the whole module but I'm not sure if its imported automatically under some circumstances. I haven't tested that.

So I'm a bit stuck.